model/riscv_insts_base.sail

```diff
@@ -313,14 +313,17 @@ function process_load(rd, vaddr, value, is_unsigned) =
     MemException(e)  => { handle_mem_exception(vaddr, e); RETIRE_FAIL }
   }
 
+function is_aligned(vaddr : xlenbits, width : word_width) -> bool =
+  match width {
+    BYTE   => true,
+    HALF   => vaddr[0..0] == zeros(),
```

Comment:
I think the vaddr[0] should be correct instead of [0..0]

Reply:
I assume the problem is bit and bits(1) are not the same thing, and this keeps it consistent with the other cases.

Reply:
Yeah exactly. Ideally I would do this:

```
function is_aligned(vaddr : xlenbits, width : word_width) -> bool =
   vaddr[word_width_bytes(word_width) >.. 0] == zeros()
```

However `>..` isn't supported yet - see [this issue](https://github.com/rems-project/sail/issues/471).

It definitely looks a bit weird to have `[N .. N]`. As well as not being able to do `bitone == 0b1` you also can't do `bitone @ 0b` which leads to weird looking code like this:

```
sign_extend(~(addr[38 .. 38]) @ addr[38 .. 0])
```

Sail does actually support overloading the operators so you can then do this:

```
function prepend_bit forall 'n, 'n >= 1. (b : bit, rest : bits('n)) -> bits('n + 1) =
  (if b == bitone then 0b1 else 0b0) @ rest

function append_bit forall 'n, 'n >= 1. (rest : bits('n), b : bit) -> bits('n + 1) =
  rest @ (if b == bitone then 0b1 else 0b0)

function join_bits(b0 : bit, b1 : bit) -> bits(2) =
  (if b0 == bitone then 0b1 else 0b0) @ (if b1 == bitone then 0b1 else 0b0)

overload append = {prepend_bit, append_bit, join_bits}
```

and then this works:

```
sign_extend(~(addr[38]) @ addr[38 .. 0])
```

which I think from a readability point of view is a good improvement, but I'm not sure about type safety. It's probably fine.

Reply:
At some point I hope to make `bit` and `bits(1)` the same type. I've almost reached the point where that's possible.

In the past (a long time ago) we just had a polymorphic vector type, and a bitvector would be a vector containing bits `vector('n, dec, bit)`. That turned out to be a mistake, as it made code generation very tricky - we nearly always want to represent bitvectors differently to other vectors. To fix that I split the type into a `vector` and `bitvector` type. However there was also the order parameter that determined the indexing order. This stopped me from being able to just make `bit` into `bits(1)` because actually behind the scenes we would have either `bitvector(inc, 1)` and `bitvector(dec, 1)`. I have subsequently removed the order parameter so now it should be possible...

Reply:
Ah interesting! I was under the impression that `bits(3)` was still the same as `vector(4, bit)`... but now that I reread the docs you do explicitly say they are different.

Making `bit` and `bits(1)` the same in Sail would be great. I guess that would mean you can eliminate `bitone` and `bitzero` too? In fact... do you need `bit` at all?